### PR TITLE
Rewritten how bars work to fix current issues, added new mistlands resource bar

### LINKED ISSUE #31: Error when enabling customfoodbar button in Config Manager
Error when enabling customfoodbar button in config manager

[Error  : Unity Log] NullReferenceException: Object reference not set to an instance of an object
Stack trace:
BetterUI.Patches.CustomElements+FoodBar.Update (Player player) (at <ba9fa52f025a4f1aab3fa5a0598003a4>:0)
BetterUI.GameClasses.BetterHud.UpdateFood (Player player) (at <ba9fa52f025a4f1aab3fa5a0598003a4>:0)
(wrapper dynamic-method) Hud.DMD<Hud::UpdateFood>(Hud,Player)
(wrapper dynamic-method) Hud.DMD<Hud::Update>(Hud)

### PR DESCRIPTION
**Pull request by Goldenrevolver**
Fixes #31

Issues that I want to address with this PR:
- user enable/ disable the health, stamina, food and xp bar ingame, this either leads to potential null reference exceptions either from the mod talking to a bar that doesn't exist or turns an existing bar non functional without reenabling the default one
- users are constantly confused that they need to restart (or how to restart) after enabling / disabling a bar, because the config description is too hidden
- the mod doesn't support resource bar for the new mistlands resource Eitr yet (I'm avoiding spoilers here since I know you haven't played yet)
- unbeknownst to the user, the new bar config options only do something when 'enablePlayerHudEditing' is also enabled
- the default display of the custom bars is not satisfying because their positions heavily overlap due to being the default instantiated positions
- users may be confused about the name of the rotation config options not specifying that they only affect the custom bars
- enabling the custom food bar while having the custom health bar disabled would lead to two food bars since we disable the entire health and food bar object when enabling the custom health bar

This PR addresses these issues in the following ways:
- after spawning or not spawning the bars in HUD.Awake based on the config, the mod completely ignores the config values and also never talks to the bar objects directly until the user logs out. That means that the bar classes have complete control over the bars after they were spawned and get to continue to function until they are destroyed by a logout

- the config file has been edited: the bar settings now have their own section that explicitely mentions that you need to logout (full restarts also work of course, but the mod only requires HUD.Awake for a full reset). this change will reset the affected config options (and only those) for ALL users, since the config options are stored as a pair of section and name. this is okay, because hitting boolean toggles is not a lot of work and most users are probably not using the custom bars in the first place. the customized positions of bars are saved in '"9 - xDataUI","uiData"' and are therefore not affected. we use this reset opportunity to also rename the rotation config options to avoid potential confusion

- the CustomElements class has been heavily modified. the three bar types shared a lot of copy pasted code. this has been refactored into a helper class which increases future maintainability. inheritance was not an option due to the heavy use of the static modifier in the existing implementation

- the fork icon next to the custom food bar has been removed again. The issue was caused by its name changing from "foodicon" to "foodicon (1)" in a recent update (probably when they removed the food bars)
- the custom food bar now does not spawn when the custom health bar is disabled and a warning is printed into the log to inform the user. additionally the descriptions of the custom health and food bar were updated to tell the user about this dependency. hopefully not spawning the bar instead of spawning two bars gets the user to notice their mistake more easily
- the Update method of the FoodBar has been updated to reflect HUD.UpdateFood again. Food icons are now blinking again when they can be refreshed

- the newly created framework has been used to add support for the resource bar of the new mistlands resource Eitr. as I have not used Eitr much I don't know if this new bar may miss out on some special visual effects (it is once again a clone of the health bar after all), but it's fully functional including when the maximum of the resource gets reset to zero
- the new bars have satisfying default positions (see the attached image), including when not all of them are enabled, and no longer require 'enablePlayerHudEditing' to be enabled to function. the default positions don't affect users with already edited positions because those are loaded afterwards and override these default positions

PRfood

### CODE REVIEW

File: BetterUI/Main.cs [hunk 1]
```diff
@@ -98,146 +100,152 @@ public Main()
...
+
+            enablePlayerHudEditing = Config.Bind(sectionName, nameof(enablePlayerHudEditing), true, "Enable the ability to edit the Player HUD by pressing a hotkey.");
+
+            useCustomHealthBar = Config.Bind(sectionName, nameof(useCustomHealthBar), false, $"Resizable, rotatable HP bar. This bar will always be the same size and will not scale when you eat. Will also disable the default food bar, so use {nameof(useCustomFoodBar)}.");
             
-            useCustomFoodBar = Config.Bind("1 - Player HUD", nameof(useCustomFoodBar), false, "Resizable, rotatable Food Bar. (REQUIRES RESTART)");
+            useCustomStaminaBar = Config.Bind(sectionName, nameof(useCustomStaminaBar), false, "Resizable, rotatable Stamina bar. This bar will always be visible and will not scale when you eat.");
 
-            healthBarRotation = Config.Bind("1 - Player HUD", nameof(healthBarRotation), 0, "Rotate healthbar in degrees  (REQUIRES RESTART)");
+            useCustomSpoilerBar = Config.Bind(sectionName, nameof(useCustomSpoilerBar), false, "Resizable, rotatable bar for the new spoiler resource. This bar will always be visible and will not scale when you eat.");
```
Comment (thedefside) [inline BetterUI/Main.cs:131]: To avoid confusion, I think this should be named what it actually is instead of 'spoiler'

File: BetterUI/Main.cs [hunk 2]
```diff
@@ -98,146 +100,152 @@ public Main()
...
+            showXPNotifications = Config.Bind(sectionName, nameof(showXPNotifications), true, "Show whenever you gain xp from actions.");
 
-            showXPNotifications = Config.Bind("5 - Character XP", nameof(showXPNotifications), true, "Show whenever you gain xp from actions.");
+            notificationTextSizeXP = Config.Bind(sectionName, nameof(notificationTextSizeXP), 14, "XP notification font size.");
 
-            notificationTextSizeXP = Config.Bind("5 - Character XP", nameof(notificationTextSizeXP), 14, "XP notification font size.");
+            extendedXPNotification = Config.Bind(sectionName, nameof(extendedXPNotification), true, "Extend notification with: (xp gained) [current/overall xp].");
 
-            extendedXPNotification = Config.Bind("5 - Character XP", nameof(extendedXPNotification), true, "Extend notification with: (xp gained) [current/overall xp]");
+            // Character XP RESTART
```
Comment (thedefside) [inline BetterUI/Main.cs:193]: Wouldn't it be better to keep this in the section it was in and add the "(Requires Logout)" in the setting description?

File: BetterUI/Main.cs [hunk 3]
```diff
@@ -98,146 +100,152 @@ public Main()
...
+            modKeySecondary = Config.Bind(sectionName, nameof(modKeySecondary), KeyCode.LeftControl,
+                "Key needed to hold down to change element dimensions. Accepted Values: https://docs.unity3d.com/ScriptReference/KeyCode.html");
 
-            useCustomHealthBar = Config.Bind("1 - Player HUD", nameof(useCustomHealthBar), false, "Resizable, rotatable HP bar. This bar will always be the same size and will not scale when you eat. (REQUIRES RESTART)");
-            
-            useCustomStaminaBar = Config.Bind("1 - Player HUD", nameof(useCustomStaminaBar), false, "Resizable, rotatable Stamina bar. This bar will always be visible and will not scale when you eat. (REQUIRES RESTART)");
+            // Player HUD RESTART
+            sectionName = "1 - Player HUD (Requires Logout)";
+
+            enablePlayerHudEditing = Config.Bind(sectionName, nameof(enablePlayerHudEditing), true, "Enable the ability to edit the Player HUD by pressing a hotkey.");
```
Comment (thedefside) [inline BetterUI/Main.cs:125]: Isn't having the (REQUIRES RESTART) in the setting description more visible than having it once in the section name? Not that people read it....

File: BetterUI/Main.cs [hunk 4]
```diff
@@ -33,10 +33,12 @@ public const string
         public static ConfigEntry<KeyCode> modKeySecondary;
         public static ConfigEntry<bool> useCustomHealthBar;
         public static ConfigEntry<bool> useCustomStaminaBar;
+        public static ConfigEntry<bool> useCustomSpoilerBar;
         public static ConfigEntry<bool> useCustomFoodBar;
-        public static ConfigEntry<int> healthBarRotation;
-        public static ConfigEntry<int> staminaBarRotation;
-        public static ConfigEntry<int> foodBarRotation;
+        public static ConfigEntry<int> customHealthBarRotation;
```
Comment (thedefside) [inline BetterUI/Main.cs:38]: I get what you're trying to do here, but is it worth resetting the current values everyone has set?

Comment (Goldenrevolver) [reply BetterUI/Main.cs:38]: as mentioned, changing these to a different section would reset the settings anyway. if we don't move them to a different section, then of course we shouldn't rename the rotation config options. but I thought if we reset them with the section change anyway, then now is a good name to improve the name

File: BetterUI/Main.cs (hunk 3 above)
Comment (Goldenrevolver) [reply BetterUI/Main.cs:125]: the setting description is a text that appears when you hover on the config option while the mouse isn't controlling the player (like in the main menu or the pause menu or inside a container). it does not appear when hovering on the toggle itself, only the label. when you change a setting ingame without 'freeing' the mouse, then you don't get to see the description at all. for those reasons I think the description is the worst place to put this information. but of course... not that people read it. it's just an idea. in my personal opinion, the section label is a bit harder to miss

File: BetterUI/Main.cs (hunk 1 above)
Comment (Goldenrevolver) [reply BetterUI/Main.cs:131]: sure, we can do that. I don't know when in progression people start modding. the existence of Eitr is the big thing mistlands is about, so I thought a person in the swamps didn't want to know about that, but in the end... who cares, we can change it to Eitr if you want

File: BetterUI/Main.cs (hunk 2 above)
Comment (Goldenrevolver) [reply BetterUI/Main.cs:193]: see below for my reasoning

File: BetterUI/Main.cs (hunk 1 above)
Comment (Goldenrevolver) [reply BetterUI/Main.cs:131]: changed for the next PR

**Review by thedefside — APPROVED** [review]

**Pull request merged**